src/mongocxx/options/index.hpp

```diff
@@ -0,0 +1,365 @@
+    };
+
+    ///
+    /// Class representing the optional WiredTiger storage engine options for indexes.
+    ///
+    class wiredtiger_storage_options : public base_storage_options {
+       public:
+        ///
+        /// Set the WiredTiger configuration string.
+        ///
+        /// @param config_str
+        ///   The WiredTiger configuration string.
+        ///
+        void config_str(std::string config_str);
+
+        ///
+        /// The current config_str setting.
+        ///
+        /// @return The current config_str.
+        ///
+        const stdx::optional<std::string>& config_str() const;
+
+       private:
+        friend collection;
+        MONGOCXX_PRIVATE const int type() const override;
+        stdx::optional<std::string> _config_str;
+    };
+
+    index();
+
+    ///
+    /// Whether or not to build the index in the background so that building
+    /// the index does not block other database activities.
+    ///
+    /// @param background
+    ///   Whether or not to build the index in the background.
+    ///
+    /// @see https://docs.mongodb.org/manual/tutorial/build-indexes-in-the-background/
+    ///
+    void background(bool background);
+
+    ///
+    /// The current background setting.
+    ///
+    /// @return The current background.
+    ///
+    const stdx::optional<bool>& background() const;
+
+    ///
+    /// Whether or not to create a unique index so that the collection will not
+    /// accept insertion of documents where the index key or keys match an
+    /// existing value in the index.
+    ///
+    /// @param unique
+    ///   Whether or not to create a unique index.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-unique/
+    ///
+    void unique(bool unique);
+
+    ///
+    /// The current unique setting.
+    ///
+    /// @return The current unique.
+    ///
+    const stdx::optional<bool>& unique() const;
+
+    ///
+    /// The name of the index.
+    ///
+    /// @param name
+    ///   The name of the index.
+    ///
+    void name(std::string name);
+
+    ///
+    /// The current name setting.
+    ///
+    /// @return The current name.
+    ///
+    const stdx::optional<std::string>& name() const;
+
+    ///
+    /// Whether or not to create a sparse index. Sparse indexes only reference
+    /// documents with the indexed fields.
+    ///
+    /// @param sparse
+    ///   Whether or not to create a sparse index.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-sparse/
+    ///
+    void sparse(bool sparse);
+
+    ///
+    /// The current sparse setting.
+    ///
+    /// @return The current sparse.
+    ///
+    const stdx::optional<bool>& sparse() const;
+
+    ///
+    /// Optionally used only in MongoDB 3.0.0 and higher. Specifies the storage
+    /// engine options for the index.
+    ///
+    /// @warning
+    ///  The base_storage_options* instance must stay in scope until after
+    ///  the call to create_index.
```

Comment:
"the call to create_index" isn't quite right, because the lifetime of this options object isn't strictly tied to a single create_index() call.  Better to say that the storage_options object must outlive this index::options object.